### PR TITLE
package/lemon: improve

### PR DESCRIPTION
**Pull request by Sinan81**
improve package file.

### CODE REVIEW

File: var/spack/repos/builtin/packages/lemon/package.py [hunk 1]
```diff
@@ -16,3 +16,49 @@ class Lemon(CMakePackage):
...
+        spec = self.spec
+        args = []
+        args.extend(
+            [
+                "-DLEMON_ENABLE_COIN={0}".format("YES" if "+coin" in spec else "NO"),
+                "-DLEMON_ENABLE_ILOG={0}".format("YES" if "+ilog" in spec else "NO"),
+                "-DLEMON_ENABLE_GLPK={0}".format("YES" if "+glpk" in spec else "NO"),
+                "-DLEMON_ENABLE_SOPLEX={0}".format("YES" if "+soplex" in spec else "NO"),
+                "-DLEMON_DEFAULT_LP={0}".format(spec.variants["lp_backend"].value.upper()),
+                "-DLEMON_DEFAULT_MIP={0}".format(spec.variants["mip_backend"].value.upper()),
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/lemon/package.py]: There are helper methods for `CMakePackage` (see https://spack.readthedocs.io/en/latest/build_systems/cmakepackage.html#adding-flags-to-cmake) though apparently not providing the expected values.

We are moving to increased use of f-strings (https://docs.python.org/3/tutorial/inputoutput.html#fancier-output-formatting) over `format()`.

Finally, we are also moving toward use of `spec.satisfies()` for checking boolean variant values (see https://spack.readthedocs.io/en/latest/packaging_guide.html#testing-spec-constraints, especially the note in the last paragraph where use of `in spec` will include variant values of dependencies.

File: var/spack/repos/builtin/packages/lemon/package.py [hunk 2]
```diff
@@ -16,3 +16,49 @@ class Lemon(CMakePackage):
...
+
+    depends_on("glpk", when="+glpk")
+    # LP solver
+    depends_on("glpk", when="lp_backend=glpk")
+    depends_on("cplex", when="lp_backend=cplex")
+    depends_on("clp", when="lp_backend=clp")
+    # MIP solver
+    depends_on("glpk", when="mip_backend=glpk")
+    depends_on("cplex", when="mip_backend=cplex")
+    depends_on("cbc", when="mip_backend=cbc")
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/lemon/package.py]: Does the software build support any combination of `glpk`, `lp_backend`, and `mip_backend`?  

If only one can be provided, then you could use `conflict()` directives (see https://spack.readthedocs.io/en/latest/packaging_guide.html#conflicts-and-requirements).

**Review by tldahlgren — COMMENTED** [review]
I have a few questions and suggestions.

File: var/spack/repos/builtin/packages/lemon/package.py (hunk 2 above)
Comment (Sinan81) [reply var/spack/repos/builtin/packages/lemon/package.py]: removed these. According to cmake files these're automatically set.
```
IF(ILOG_FOUND)
  SET(DEFAULT_LP "CPLEX")
  SET(DEFAULT_MIP "CPLEX")
ELSEIF(COIN_FOUND)
  SET(DEFAULT_LP "CLP")
  SET(DEFAULT_MIP "CBC")
ELSEIF(GLPK_FOUND)
  SET(DEFAULT_LP "GLPK")
  SET(DEFAULT_MIP "GLPK")
ELSEIF(SOPLEX_FOUND)
  SET(DEFAULT_LP "SOPLEX")
ENDIF()
```

File: var/spack/repos/builtin/packages/lemon/package.py (hunk 1 above)
Comment (Sinan81) [reply var/spack/repos/builtin/packages/lemon/package.py]: Thanks for the suggestions. So, should I do the following instead:
```python
    def cmake_args(self):
        spec = self.spec
        args = []
        args.extend(
            [
                f"-DLEMON_ENABLE_COIN={spec.variants['coin'].value}",
                f"-DLEMON_ENABLE_ILOG={spec.variants['ilog'].value}",
                f"-DLEMON_ENABLE_GLPK={spec.variants['glpk'].value}",
                f"-DLEMON_ENABLE_SOPLEX={spec.variants['soplex'].value}",
            ]
        )
        return args
```

Comment (Sinan81) [reply var/spack/repos/builtin/packages/lemon/package.py]: i checked that the package builds fine. It seems cmake doesn't care if it's a yes/no or true/false.

File: var/spack/repos/builtin/packages/lemon/package.py [hunk 3]
```diff
@@ -17,36 +17,20 @@ class Lemon(CMakePackage):
...
-    # LP solver
-    depends_on("glpk", when="lp_backend=glpk")
-    depends_on("cplex", when="lp_backend=cplex")
-    depends_on("clp", when="lp_backend=clp")
-    # MIP solver
-    depends_on("glpk", when="mip_backend=glpk")
-    depends_on("cplex", when="mip_backend=cplex")
-    depends_on("cbc", when="mip_backend=cbc")
+    depends_on("cplex", when="+ilog")
+    depends_on("coinutils", when="+coin") # just a guess
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/lemon/package.py]: Do you know how to confirm whether this is the right package?

File: var/spack/repos/builtin/packages/lemon/package.py [hunk 4]
```diff
@@ -17,36 +17,20 @@ class Lemon(CMakePackage):
...
+    depends_on("cbc", when="+coin")
+    depends_on("clp", when="+coin")
+    depends_on("soplex", when="+soplex") # no such package in Spack yet.
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/lemon/package.py]: When do you plan to add this package?

**Review by tldahlgren — COMMENTED** [review]
Some questions.

File: var/spack/repos/builtin/packages/lemon/package.py (hunk 4 above)
Comment (tldahlgren) [reply var/spack/repos/builtin/packages/lemon/package.py]: It needs to be in place before this PR can be merged.

Comment (Sinan81) [reply var/spack/repos/builtin/packages/lemon/package.py]: I'll comment it out and leave it as TODO.

File: var/spack/repos/builtin/packages/lemon/package.py (hunk 3 above)
Comment (Sinan81) [reply var/spack/repos/builtin/packages/lemon/package.py]: It's seems they're strictly related.

Comment (Sinan81) [reply var/spack/repos/builtin/packages/lemon/package.py]: it's a direct dependency of `cbc` package, also referred in this packagefile.

Comment (Sinan81) [reply var/spack/repos/builtin/packages/lemon/package.py]: `+coin` variant is currently not building due to error:
```
  >> 63    CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
     64    Please set them or make sure they are set and tested correctly in the CMake files:
     65    COIN_BZ2_LIBRARY (ADVANCED)
     66        linked by target "lemon" in directory /cache/sbulut/spack-stage/spack-stage-lemon-1.3.1-cbkh7hrng7fqhfrcoothdx5ezkqeig4c/spack-src/lemon
     67        linked by target "lemon" in directory /cache/sbulut/spack-stage/spack-stage-lemon-1.3.1-cbkh7hrng7fqhfrcoothdx5ezkqeig4c/spack-src/lemon
     68        linked by target "lp_test" in directory /cache/sbulut/spack-stage/spack-stage-lemon-1.3.1-cbkh7hrng7fqhfrcoothdx5ezkqeig4c/spack-src/test
     69        linked by target "mip_test" in directory /cache/sbulut/spack-stage/spack-stage-lemon-1.3.1-cbkh7hrng7fqhfrcoothdx5ezkqeig4c/spack-src/test
```

Comment (Sinan81) [reply var/spack/repos/builtin/packages/lemon/package.py]: I am only interested in `+glpk` variant really. We can probably leave this as TODO.

**Comment by Sinan81** [comment]
@spackbot fix style

**Review by tldahlgren — APPROVED** [review]

**Pull request merged**